src/Build/Evaluation/Expander.cs

```diff
@@ -2022,9 +2058,28 @@ ExpanderOptions options
                     return true;
                 }
 
+                int startLength = builder.Length;
+                bool truncate = IsTruncationEnabled(options);
+
                 // if the capture.Separator is not null, then ExpandExpressionCapture would have joined the items using that separator itself
-                foreach (var item in itemsFromCapture)
+                for (int i = 0; i < itemsFromCapture.Count; i++)
                 {
+                    var item = itemsFromCapture[i];
+                    if (truncate)
+                    {
+                        if (i >= ItemLimitPerExpansion)
+                        {
+                            builder.Append("...");
+                            return false;
+                        }
+                        int currentLength = builder.Length - startLength;
+                        if (currentLength + item.Key.Length > CharacterLimitPerExpansion)
+                        {
+                            builder.Append(item.Key.Substring(0, CharacterLimitPerExpansion - currentLength - 3));
```

Comment:
Is it not possible for the length of the substring to be negative? Say that:
```C#
itemsFromCapture[0].Key.Length == CharacterLimitPerExpansion
```
so we get to this line for index 1 and `CharacterLimitPerExpansion - currentLength - 3` could be as bad as -4 because `currentLength` could be as high `CharacterLimitPerExpansion  + 1` (the 1 is for the semicolon).

Reply:
Good catch. Should it substring to `min(item.key.length, max(0, CharacterLimitPerExpansion - currentLength - 3)`) . The max is to ensure it does not go bellow 0, the min is to sanity ensure it does not go over max string length. I did not factor in the off by ones by the semicolons.

Reply:
Yeah, I think I have to check if this value is > 0 and not append anything at all in this case.

I'll check the other places `- 3` is used.

Reply:
You can append something as long as you have a stricter check, right?

`if (currentLength + item.Key.Length + 3 > CharacterLimitPerExpansion)`
...
?

Reply:
I updated the test case so it caused: 
```
    System.ArgumentOutOfRangeException : Length cannot be less than zero.
    Parameter name: length
```
Then fixed the problem, thanks. 👍 

The other places do:
```csharp
Substring(0, CharacterLimitPerExpansion - 3) + "..."
```
This should be fine as `1024 - 3`.